src/components/ExperimentsOverview.tsx

```
@@ -856,6 +868,19 @@ function PromotedTraffic({ enabled }: { enabled: boolean }) {
     [filteredPoints, showCulled, isPromotedCulled],
   );
 
+  // Paginate BEFORE byMarket: the SigNoz fan-out derives from the visible page,
+  // so paging also caps the per-render query volume (same reason culled-hiding
+  // happens before byMarket). Page resets whenever the filtered universe
+  // changes; the clamp additionally covers shrinkage from async liveness data.
+  const [page, setPage] = useState(0);
+  useEffect(() => { setPage(0); }, [country, protocol, provider, showCulled, hours]);
```

ESLint

[error] Error: Calling setState synchronously within an effect can trigger cascading renders

Effects are intended to synchronize state between React and external systems such as manually updating the DOM, state management libraries, or other platform APIs. In general, the body of an effect should do one or both of the following:
* Update external systems with the latest state from React.
* Subscribe for updates from some external system, calling setState in a callback function when external state changes.

Calling setState synchronously within an effect body causes cascading renders that can hurt performance, and is not recommended. (https://react.dev/learn/you-might-not-need-an-effect).

/home/jailuser/git/src/components/ExperimentsOverview.tsx:876:21
  874 |   // changes; the clamp additionally covers shrinkage from async liveness data.
  875 |   const [page, setPage] = useState(0);
> 876 |   useEffect(() => { setPage(0); }, [country, protocol, provider, showCulled, hours]);
      |                     ^^^^^^^ Avoid calling setState() directly within an effect
  877 |   const pageCount = Math.max(1, Math.ceil(promotions.length / PROMOTIONS_PAGE_SIZE));
  878 |   const clampedPage = Math.min(page, pageCount - 1);
  879 |   const pagePromotions = useMemo(

(react-hooks/set-state-in-effect)